users/tests/test_unit_users_apiview.py

```diff
@@ -0,0 +1,236 @@
+        users = list(dictionary.values())[0]
+        self.assertTrue(self.user in users)
+
+    def test_user_list_view_should_not_be_accessible_for_unauthenticated_user(self):
+        request = APIRequestFactory().get(path=self.url)
+        request.user = AnonymousUser()
+        response = views.UserList.as_view()(request)
+        self.assertEqual(response.status_code, 403)
+
+
+class UserDetailTests(TestCase):
+    def setUp(self):
+        self.user = CustomUser(
+            email="testuser@codepoets.it",
+            password='newuserpasswd',
+            first_name='John',
+            last_name='Doe',
+            country='PL'
+        )
+        self.user.full_clean()
+        self.user.save()
+
+    def test_user_detail_view_should_display_user_details_on_get(self):
+        request = APIRequestFactory().get(path=reverse('custom-users-detail', args=(self.user.pk,)))
+        request.user = self.user
+        response = views.UserDetail.as_view()(request, pk=self.user.pk)
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, self.user.email)
+        self.assertEqual(response.data['serializer'].instance, self.user)
+
+    def test_user_detail_view_should_not_render_non_existing_user(self):
+        not_existing_pk = 1000
+        request = APIRequestFactory().get(path=reverse('custom-users-detail', args=(not_existing_pk,)))
+        request.user = self.user
+        response = views.UserDetail.as_view()(request, pk=not_existing_pk)
+        self.assertEqual(response.status_code, 404)
+
+    def test_user_detail_view_should_update_user_on_post(self):
+        old_phone_number = ''.join(random.choices(string.digits, k=constants.PHONE_NUMBER_MIN_LENGTH))
+        response = self.client.post(
+            path=reverse('custom-user-update-by-admin', args=(self.user.pk,)),
+            data={
+                'email': self.user.email,
+                'phone_number': old_phone_number,
+            },
+        )
+        self.user.refresh_from_db()
+        current_phone_number = self.user.phone_number
+        self.assertEqual(response.status_code, 302)
+        self.assertEqual(current_phone_number, old_phone_number)
+
+    def test_user_detail_view_should_not_update_user_on_post_if_form_is_invalid(self):
+        old_phone_number = ''.join(random.choices(string.digits, k=constants.PHONE_NUMBER_MIN_LENGTH - 1))
+        request = APIRequestFactory().post(
+            path=reverse('custom-user-update-by-admin', args=(self.user.pk,)),
+            data={
+                'email': self.user.email,
+                'phone_number': old_phone_number,
+            },
+        )
+        response = views.UserUpdateByAdmin.as_view()(request, pk=self.user.pk)
+        self.user.refresh_from_db()
+        current_phone_number = self.user.phone_number
+        self.assertEqual(response.status_code, 200)
+        self.assertNotEqual(current_phone_number, old_phone_number)
```

Comment:
You shouldn't check if something is not equal, of course in some cases it is good to check if something != something, but here we should check if phone number didn't change. So please at the test beginning please extract phone number and at the end -> ```self.assertEqual(number_before_request, self.user.phone_number)```